### PR TITLE
USHIFT-1083 Back up data if previous boot was healthy

### PR DESCRIPTION
**Pull request by pmtk**

### CODE REVIEW

File: pkg/admin/prerun/advisor_backup.go [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+	return a.data.BootHistory != nil
+}
+
+func (a *backupAdvisor) BootHistoryContainsPreviousBoot() bool {
+	return a.data.PreviousBootInfo != nil
+}
+
+func (a *backupAdvisor) PreviousBootWasHealthy() bool {
+	if !a.BootHistoryContainsPreviousBoot() {
+		panic("PreviousBootWasHealthy() should not be called if BootHistoryContainsPreviousBoot() returned false")
```
Comment (dhellmann) [inline pkg/admin/prerun/advisor_backup.go:39]: Could this function just return false in this case, to make it safer to use?

**Comment by pmtk** [comment]
@dhellmann  I moved RF test to https://github.com/openshift/microshift/pull/1866

File: packaging/greenboot/green.sh [hunk 2]
```diff
@@ -0,0 +1,3 @@
+#!/usr/bin/env bash
+
+microshift admin ostree set-health healthy
```
[the remark this answers is not in the document]
Comment (pmtk) [reply packaging/greenboot/green.sh:3]: Initially there was supposed to be `prerun` subcommand, but I incorporated into `run` so I'll drop one level

[the remark this answers is not in the document]
Comment (pmtk) [reply packaging/greenboot/green.sh]: I made more changes: removed `admin` command, `data` and `health` are top level commands, `health` is hidden (not shown in help)

File: packaging/greenboot/set-health.sh [hunk 3]
```diff
@@ -0,0 +1,13 @@
+#!/usr/bin/env bash
+
+SCRIPT_DIR="$(readlink -f "$(dirname "${BASH_SOURCE[0]}")")"
+
+if [[ "${SCRIPT_DIR}" == *"green.d"* ]]; then
```
Comment (dhellmann) [inline packaging/greenboot/set-health.sh:5]: This feels tricky. Could we just have 2 scripts so we're explicitly passing the health status?

**Review by dhellmann — COMMENTED** [review]
This PR is growing pretty big, and it's doing a lot of things that don't seem to be in scope for the description of the original ticket.

Could we store the health state as a simple string in a file by itself? That would let the greenboot integration just be shell scripts that write to the file, and then MicroShift could read that file to check the state on startup and decide whether to do the backup. Then we could bring some of the rest of the accounting stuff for history in via separate PRs linked to some of the other tickets, where the code changes would be smaller and easier to review.

**Comment by pmtk** [comment]
I'll split this PR into smaller pieces starting with https://github.com/openshift/microshift/pull/1874
/close

**Pull request closed without merging**